pcie: Add runtime lookup of BDF values

jhedberg opened the pull request:
The BDF values can differ on the same platform, based on e.g. BIOS
configuration, and in the case of qemu the command line parameters. It's
therefore more reliable to always look up the BDF value based on the
known Vendor and Device IDs.

Fixes #26109

Commit:
pcie: Add support for centralized lookup of BDF values

The BDF values can differ on the same platform, based on e.g. BIOS
configuration, and in the case of qemu the command line parameters. It's
therefore more reliable to always look up the BDF value based on the
known Vendor and Device IDs.

This patch introduces such a framework, and allows the incremental
update of PCIe drivers to start taking advantage of it.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
drivers: uart_ns16550: Convert to use runtime PCIe BDF lookup

Convert the ns16550 driver to use the new centralized runtime BDF lookup
of PCIe devices.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
drivers: eth_e1000: Convert to use dynamic BDF lookup

Use the new PCIe core infrastructure for looking up the BDF at runtime
based on the VID/DID values.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

finikorg commented on include/zephyr/drivers/pcie/pcie.h:
```diff
@@ -36,6 +37,23 @@ typedef uint32_t pcie_bdf_t;
+
+#define DEVICE_PCIE_DECLARE(node_id)                                        \
+	STRUCT_SECTION_ITERABLE(pcie_dev, Z_DEVICE_PCIE_NAME(node_id)) = {  \
+		.id = PCIE_DT_ID(node_id),                                  \
```
We need to set also `.bdf = PCIE_BDF_NONE`

jhedberg replied:
Thanks. Fixed.

finikorg commented on drivers/pcie/host/pcie.c:
```diff
@@ -368,3 +369,48 @@ pcie_bdf_t pcie_bdf_lookup(pcie_id_t id)
+				}
+
+				STRUCT_SECTION_FOREACH(pcie_dev, dev) {
+					if (dev->bdf != PCIE_BDF_NONE) {
```
works after bdf set

finikorg commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -317,7 +316,11 @@ static void e1000_iface_init(struct net_if *iface)
 }
 
-static struct e1000_dev e1000_dev;
+DEVICE_PCIE_DECLARE(0);
```
we need to provide node_id instead on instance idx

finikorg commented on include/zephyr/drivers/pcie/pcie.h:
```diff
@@ -36,6 +37,23 @@ typedef uint32_t pcie_bdf_t;
+
+#define Z_DEVICE_PCIE_NAME(node_id) _CONCAT(pcie_dev_, DT_DEP_ORD(node_id))
+
+#define PCIE_DT_ID(node_id) PCIE_ID(DT_PROP_OR(node_id, vendor_id, 0xffff), \
```
Probably simple comment / API documentation would not hurt.

jhedberg replied:
Agreed. I'll add that before promoting to normal PR from draft.

finikorg commented on samples/hello_world/src/main.c:
```diff
@@ -5,8 +5,23 @@
+	STRUCT_SECTION_COUNT(pcie_dev, &dev_count);
+	printk("%d\n", dev_count);
+
+	STRUCT_SECTION_FOREACH(pcie_dev, dev) {
```
Can some of this be added to tests (if there is appropriate place)

jhedberg replied:
I didn't intend to push this commit. Removed now.

jhedberg commented on drivers/can/can_kvaser_pci.c:
```diff
@@ -172,7 +169,10 @@ const struct can_driver_api can_kvaser_pci_driver_api = {
                                                                                                    \
-	static struct can_kvaser_pci_config can_kvaser_pci_data_##inst;                            \
+	DEVICE_PCIE_INST_DECLARE(inst);                                                            \
+	static struct can_kvaser_pci_data can_kvaser_pci_data_##inst = {                           \
```
@henrikbrixandersen note that this is also fixing a bug. The struct type was wrong, i.e. is should always have been "data" and not "config". This dropped between the cracks since it gets assigned to a void pointer and then in the functions back to the expected type. I guess the structs were close enough in size that there was no noticeable breakage, but there was a ticking time bomb here..

henrikbrixandersen replied:
Oops, thanks.

henrikbrixandersen replied:
@jhedberg Care to submit that bug fix as a separate PR since this one is still in draft?

jhedberg replied:
@henrikbrixandersen done: #52170

Commit:
drivers: can: kvaser_pci: Convert to use dynamic BDF lookup

Use the new PCIe core infrastructure for looking up the BDF at runtime
based on the VID/DID values.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

henrikbrixandersen commented:
@jhedberg Regarding the differences in PCIe BDF values due to QEMU command line changes - perhaps we ought to unconditionally add the various PCI(e) hardware described in the dts via QEMU command lines specified at board level? This doesn't solve the BIOS configuration issue, but it would make the QEMU instances unconditionally match their dts.

Commit:
drivers: i2c_dw: Convert to use dynamic BDF lookup

Use the new PCIe core infrastructure for looking up the BDF at runtime
based on the VID/DID values.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
> @jhedberg Regarding the differences in PCIe BDF values due to QEMU command line changes - perhaps we ought to unconditionally add the various PCI(e) hardware described in the dts via QEMU command lines specified at board level? This doesn't solve the BIOS configuration issue, but it would make the QEMU instances unconditionally match their dts.

Yes, there are many ways that we could fight against guessing wrong the BDF at build time. However, I think it's long-term safer to just abandon this fight by simply doing the lookup at runtime. This is how essentially all Linux drivers do it, and there are many more drivers and platforms there, existing for many more years than Zephyr, so I think it's safe to assume that it's a good approach. One of the reasons this had been avoided in Zephyr was the lack of centralised PCIe bus management, which would have required all drivers to scan for their devices independently, and that would have an unacceptable overhead. With this PR the scan is done a single time, basically same as Linux, except Linux does probe() callbacks whereas we just store the BDF into a table and let the drivers keep their existing init callbacks with whatever level & priority they want.

finikorg commented:
I tested e1000, works fine with the recent changes

finikorg commented on drivers/i2c/i2c_dw.c:
```diff
@@ -938,33 +940,39 @@ static int i2c_dw_initialize(const struct device *dev)
 #define I2C_DEVICE_INIT_DW(n)                                                 \
 	PINCTRL_DW_DEFINE(n);                                                 \
+	I2C_PCIE_DEFINE(n);                                                   \
 	static void i2c_config_##n(const struct device *port);                \
```
nitpick, port can be renamed to dev as above.

Commit:
soc: x86: Used fixed BDF values for early serial

Since DTS doesn't contain the BDF anymore (it's looked up at runtime),
hardcode the BDF value for the use of early serial.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

finikorg commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -317,7 +316,11 @@ static void e1000_iface_init(struct net_if *iface)
 }
 
-static struct e1000_dev e1000_dev;
+DEVICE_PCIE_INST_DECLARE(0);
```
We can also convert driver initialization to unreadable macrobatic code to support more then one instance.

jhedberg replied:
Sure. But I'll leave that outside of the scope of this PR. It's an enhancement that could have been done to the e1000 driver even without this PR.

Commit:
drivers: pci: ptm: Convert to use dynamic BDF lookup

Use the new PCIe core infrastructure for looking up the BDF at runtime
based on the VID/DID values.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
drivers: pci: ptm: Fix missing DT_DRV_COMPAT

The ptm.c driver never defined DT_DRV_COMPAT, so the various DT macros
never expanded to anything useful.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
drivers: virtualization: ivshmem: Convert to use dynamic BDF lookup

Use the new PCIe core infrastructure for looking up the BDF at runtime
based on the VID/DID values.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
I've now also added documentation for the new public pcie.h macros

tbursztyka reviewed (dismissed):
lgtm

finikorg commented on include/zephyr/drivers/pcie/pcie.h:
```diff
@@ -36,6 +37,95 @@ typedef uint32_t pcie_bdf_t;
+/**
+ * @brief Get the PCIe Vendor and Device ID for a node
+ *
+ * @param node_id DTS node_id
```
I would change it to `DTS node identifier`

jhedberg replied:
Fixed, thanks

finikorg reviewed (dismissed):
LGTM

enjiamai reviewed (dismissed):
It's nice and verified on up2, ehl_crb, acrn_ehl_crb, etc.

henrikbrixandersen reviewed (dismissed):
CAN part looks fine.

jhedberg commented on dts/bindings/pcie/host/pcie-device.yaml:
```diff
@@ -0,0 +1,14 @@
+properties:
+    vendor-id:
+      type: int
+      required: false
```
Fixed

jhedberg commented on dts/bindings/serial/ns16550.yaml:
```diff
@@ -2,11 +2,11 @@ description: ns16550 UART
 properties:
     reg:
-      required: true
+      required: false
```
Fixed

jhedberg commented:
I've taken into account the improvement suggestions from @finikorg and @mbolivar-nordic and pushed an update. There are no other changes. @dcpleung @henrikbrixandersen @enjiamai @tbursztyka @finikorg could you take another look and re-approve, thanks!

dcpleung reviewed (approved):
Future enhancement suggestion: instead of passing `data->pcie->bdf`, simply pass `data->pcie` (the struct).
